### PR TITLE
Add ssh-credentials support

### PR DESCRIPTION
**Pull request by zregvart**
Hi, I've cobbled the support for ssh-credentials. What I hoped for was to make it optional, so it doesn't interfere with older Jenkins installs that cannot, or simply do not use the credentials plugin. I would like to work on this further by contributing to the credentials plugin so that the credentials select Jelly taglib has the name attribute, so that the ugly workaround I employed here could be removed.

Could you please review and see if this is the way to go about adding ssh-credentials support, and if indeed I've managed to make it truly optional.

Thanks ;)

### CODE REVIEW

**Comment by kristofferpeterhansel** [comment]
Is there any reason this isn't getting merged?

**Comment by oleg-nenashev** [comment]
The code and the current implementation are not backward compatible, hence the merge requires an increment of the major version. IMO it is not a blocker for the merge, because there were many such releases devoted to the credentials support

File: src/main/java/org/jvnet/hudson/plugins/CredentialsSSHSite.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+import com.cloudbees.plugins.credentials.domains.Domain;
+import com.cloudbees.plugins.credentials.domains.DomainRequirement;
+import com.cloudbees.plugins.credentials.domains.HostnamePortRequirement;
+import com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl;
+import com.google.common.base.Strings;
+import com.jcraft.jsch.ChannelExec;
+import com.jcraft.jsch.JSchException;
+import com.jcraft.jsch.Session;
+
+public class CredentialsSSHSite {
```
Comment (oleg-nenashev) [inline src/main/java/org/jvnet/hudson/plugins/CredentialsSSHSite.java]: I would leave the original name of the class if it is possible

File: src/main/java/org/jvnet/hudson/plugins/CredentialsSSHSite.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+	public String getHostname() {
+		return hostname;
+	}
+
+	public int getIntegerPort() {
+		return port;
+	}
+
+	public String getPort() {
+		return "" + port;
```
Comment (oleg-nenashev) [inline src/main/java/org/jvnet/hudson/plugins/CredentialsSSHSite.java]: ???
It is not a good way to achieve the @Nonnull behaviour

**Comment by mfolnovic** [comment]
Any progress on merging this?

**Comment by daniel-beck** [comment]
@edmund-wagner Are you still maintaining this plugin? What's your opinion on this change? @johnny-b-goode How about you?

**Comment by johnny-b-goode** [comment]
I would like to do a release with a couple of bugfixes, possibly this weekend if time permits. Once that is done this will be the next thing I look at. The limiting factor is simply having time.

**Comment by edmund-wagner** [comment]
I'm a bit scared of the testing. will try to look at this this evening...

**Comment by daniel-beck** [comment]
@johnny-b-goode @edmund-wagner If you have concerns about the data migration, consider releasing into the experimental update center and ask users to test this on their test instances. Just use a version like `2.6-beta-1`.
http://jenkins-ci.org/content/experimental-plugins-update-center
https://github.com/jenkinsci/backend-update-center2/blob/master/src/main/java/org/jvnet/hudson/update_center/MavenArtifact.java#L155...L158

Since this changes the data format, a compatibility warning would get users to (hopefully) read the changelog before upgrading, so they know what's going on:
https://wiki.jenkins-ci.org/display/JENKINS/Marking+a+new+plugin+version+as+incompatible+with+older+versions

**Comment by orrc** [comment]
Just wondering if there's any progress on making a new release (even an experimental one)?

**Comment by edmund-wagner** [comment]
still need to figure out what goes wrong with the migration to the credentials store. only one user/host combination is migrated as of now...

**Comment by ljader** [comment]
Unfortunately, commit with the merge 5ec34f890c230b4fd35e01bd6b7cfaea8ad9d2a7 contains incorrect conflict resolution, and broke ssh connection execution.
I'm working on fixing this as well as the credentials issue.

**Pull request merged**